README.md
[![Pub Package](https://img.shields.io/pub/v/memcache.svg)](https://pub.dartlang.org/packages/memcache)
[![Build Status](https://travis-ci.org/dart-lang/memcache.svg?branch=master)](https://travis-ci.org/dart-lang/memcache)

## Memcache interface for the `appengine` package.

This package define Dart interfaces for memcache. The package defines both
a high-level interface and a low-level interface.

These interfaces are used for the memcache support in the `appengine` package.
Therefore these interfaces only support the memcache operations supported by
App Engine.

### High-level interface

The high-level interface is called `Memcache`. It provides a `Future` based API
for the most common access patterns. It defaults to use `String` values for
both keys and values. Store and retrieve single values:

```
Memcache memcache = ...
memcache.set('MyKey', 'MyValue').then((_) {
  memcache.get('MyKey').then((String value) {
    // value is read from memcache.
  });
});
```

Store and retrieve multiple values:

```
Memcache memcache = ...
memcache.setAll({'MyKey1': 'MyValue1',
                 'MyKey2': 'MyValue2').then((_) {
  memcache.getAll(['MyKey1', 'MyKey2']).then((Map result) {
    // result is a map with the values for MyKey1 and MyKey2.
  });
});
```

The high-level interface does not expose the CAS values directly. However
through the `withCAS` method, it becomes simple to use CAS with memcache
operations. `withCAS` returns an implemetation of `Memcache` which internally
keeps track of the CAS for all keys retrieved. When a key is stored the CAS
value from when it was retrieved, will be passed automatically to memcache.

```
Memcache memcache = ...
Memcache cas = memcache.withCAS();
cas.get('MyKey').then((value) {
  memcache.set('MyKey', 'MyNewValue').then((_) {
    // MyNewValue is stored in memcache if the CAS did not
    // change from when it was retrieved.
  });
});
```

### Low-level interface

The low-level interface is called `RawMemcache` and provides a generic
interface using request/response objects send in batches to memcache.

### Using these interfaces in an App Engine application.

In a Dart App Engine application using the `appengine` package the memcahce
interface is available on the client context.

    context.services.memcache.get('MyKey').then(String value) => ...

### Native protocol implementation

This package will also provide an implementation of the native memcached
protocol. This is still work in progress, and it will support the interfaces
defined in this package.


CHANGELOG.md
## 0.3.0+2

* Fix `onError` type cast error.

## 0.3.0

* Update code for Dart 2.

## 0.2.2

* Ignore exceptions when closing connection in error state.

## 0.2.1+2

* Fix bug in [Memcache.get] to throw an exception if we a get operation resulted
  in neither found/not-found but rather in an error (i.e. if the tcp connection
  is fine, but the server returns an error instead found/not-found).

## 0.2.1+1

* Added [Memcache.fromRaw] constructor.

## 0.2.1

* Added [BinaryMemcacheProtocol] which implements the [RawMemcache] API.

* Fix bugs in the [MemCacheNativeConnection] which used typed-data views
  incorrectly.

* Switch from `package:unittest` to `package:test`.

## 0.2.0

* Added expiration to the RawMemcache API when adding values. Pass the
  expiration from the Memcache implemented on RawMemcache.

* Removed the expiration from the clear method on the Memcache interface. Some
  implementations do not support it.

## 0.1.0

* First release.


tools/run_tests.sh
#!/bin/bash

REPO_ROOT=$( cd $(dirname $(dirname "${BASH_SOURCE[0]}" )) && pwd )

# Source utility functions
source "$REPO_ROOT/tools/utils.sh"

# Check that the necessary environment variables are set.
check_env_variable "DART_SDK"

export PATH="$PATH:$DART_SDK/bin"
export RETURN_VALUE=0


start_phase "Analyzing"
for testfile in $(find $REPO_ROOT/lib -name "*.dart"); do
  analyze_file "$testfile"
  RETURN_VALUE=$(expr $RETURN_VALUE + $?)
done
for testfile in $(find $REPO_ROOT/test -name '*.dart'); do
  analyze_file "$testfile"
  RETURN_VALUE=$(expr $RETURN_VALUE + $?)
done


start_phase "Testing"
for testfile in $(find $REPO_ROOT/test -name "*test.dart"); do
  test_file "$testfile"
  RETURN_VALUE=$(expr $RETURN_VALUE + $?)
done


# Wait until background jobs are done.
wait
echo
echo

exit $RETURN_VALUE


test/test_utils.dart
// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:memcache/memcache.dart';
import 'package:test/test.dart';

const isMemcacheError = const TypeMatcher<MemcacheError>();
const isMemcacheNotStored = const TypeMatcher<NotStoredError>();
const isMemcacheModifiedError = const TypeMatcher<ModifiedError>();


lib/memcache.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library memcache;

import "dart:async";

import "memcache_raw.dart";
import "src/memcache_impl.dart" show MemCacheImpl;

/**
 * General memcache exception.
 */
class MemcacheError implements Exception {
  final Status status;
  final String message;

  const MemcacheError(this.status, this.message);

  const MemcacheError.internalError()
      : message = 'An internal error occurred.',
        status = null;

  String toString() => 'MemcacheError(status: $status, message: $message)';
}

/**
 * Exception thrown by [Memcache.set] and [Memcache.setAll] if the item was
 * not stored.
 */
class NotStoredError extends MemcacheError {
  const NotStoredError() : super(null, 'Item was not stored.');
}

/**
 * Exception thrown by [Memcache.set] and [Memcache.setAll] if CAS tracking
 * is used and the item has been modified after it was fetched.
 */
class ModifiedError extends MemcacheError {
  const ModifiedError()
      : super(null, 'Item exists with a different CAS value.');
}

/**
 * Exception thrown if there was a network error.
 */
class NetworkException extends MemcacheError {
  final error;
  NetworkException(this.error) : super(null, null);

  String toString() => 'NetworkError: "$error"';
}

/**
 * Access to a memcache service.
 *
 * The memcache service – referred to as simply 'memcache' below – provides a
 * shared cache organized as a map from keys to values where both keys and
 * values are binary values (i.e. list of bytes).
 * For each entry in the cache an expiry time can be set to ensure the item is
 * evicted from the cache after a given interval or at a specific point in time.
 * The cache has a limited size, so items can be evicted by the service at any
 * point in time, typically using a LRU-policy.
 *
 * The key value used in memcache is a binary value with a maximum length of
 * 250 bytes.
 *
 * The values stored in memcache are binary values. The maximum size for a
 * value depends on the configuration of the memcache service. The most common
 * default is 1M (one megabyte).
 *
 * In all cases where a key is passed it can be of type `List<int>` or of
 * type `String`. In the case of `List<int>` the key value is passed directly
 * to memcache. In the case of a `String` the key value is converted into
 * bytes by using the UTF-8 encoding. Note that when combining both types of
 * keys in an application a key of one type can alias a key of the other type.
 * E.g. the keys `[64]` and `'A'` are the same key.
 *
 * When the type `List<int>` is used note the values are bytes and each value
 * in the list must be in the range [0..255]. Using the class `Uint8List`
 * from `dart:typed_data` provides a compact and efficient represetation.
 */
abstract class Memcache {
  /**
   * Makes a new [Memcache] instance based on a [RawMemcache].
   */
  factory Memcache.fromRaw(RawMemcache raw) => new MemCacheImpl(raw);

  /**
   * Retreives a value from memcache.
   *
   * If the value is not found the future completes with `null`.
   *
   * If the value is found the value of [asBinary] determins the type of
   * the value the future completes with. If [asBinary] is `false` (the
   * default) the future completes with a `String`. Otherwise the future
   * completes with a list of bytes.
   *
   * The internal representation in memcache is binary. When a `String` is
   * returned it uses a UTF-8 decoder to produce a `String`.
   */
  Future get(key, {bool asBinary: false});

  /**
   * Retreives multiple values from memcache.
   *
   * The values are returned in a map where the keys are the same instances
   * as where passed in the [keys] argument.
   */
  Future<Map> getAll(Iterable keys, {bool asBinary: false});

  /**
   * Sets the value for a key in memcache. The value is set unconditionally.
   *
   * If [expiration] is not set the value is set without
   * any explicit lifetime. [expiration] cannot exceed 30 days.
   *
   * If [action] is [SetAction.SET] (the default) the
   * value is set unconditionally in memcache. That is, if the key does not
   * already exist it is created and if the key already exists its current
   * value is overwritten.
   * The value [SetAction.ADD] is used to indicate that the value will only be
   * set if the key was not in mamcache already. The value
   * [SetAction.REPLACE] is used to indicate that the value will only be set
   * if the key was already in memcache.
   *
   * The key and value can have type `String` or `List<int>`. When a `String`
   * is used for either key or value a UTF-8 encoder is used to produce the
   * binary value which is stored in memecache.
   *
   *     Memcache m = ...;
   *     m.set('mykey', 'myvalue');
   *     m.set([0, 1, 2], [3, 4, 5], expiration: new Duration(hours: 1));
   */
  Future set(key, value,
      {Duration expiration, SetAction action: SetAction.SET});

  /**
   * Sets multiple values in memcache.
   *
   * The keys and values passed in `keysAndValues` can have type `String`
   * or `List<int>`.
   *
   *     Memcache m = ...;
   *     m.setAll(
   *         {'mykey': 'myvalue',
   *          [0, 1, 2]: [3, 4, 5]
   *         });
   *
   * The values passed for [expiration] and [action] are applied to all values
   * set.
   *
   * See [set] for information on [expiration] and [action].
   *
   * Note that memcache is not transactional, so if this operation fails some
   * of the updates might still have succeeded.
   *
   * Also note that when this operation fails the exception reported is the
   * first failure encountered. There might be different failures as well.
   */
  Future setAll(Map keysAndValues,
      {Duration expiration, SetAction action: SetAction.SET});

  /**
   * Removes the key from memcache.
   */
  Future remove(key);

  /**
   * Removes all the keys in [keys] from memcache.
   *
   * Note that memcache is not transactional, so if this operation fails some
   * of the keys might still have been removed.
   */
  Future removeAll(Iterable keys);

  /**
   * Increments the value of a key in memcache.
   *
   * The current value of [key] will be incremented with the value [delta]. If
   * [key] does not exist it will be added with the value [initialValue].
   *
   * Increment in memcache works on 64-bit unsigned integers. If incrementing
   * causes the value to exceed the maximum value for a 64-bit integer the
   * value will wrap.
   *
   * If the value is negative this will work as if `decrement` was called.
   *
   * The updated value is returned.
   */
  Future<int> increment(key, {int delta: 1, int initialValue: 0});

  /**
   * Decrement the value of a key in memcache.
   *
   * The current value of [key] will be incremented with the value [delta]. If
   * [key] does not exist it will be added with the value [initialValue].
   *
   * Decrement in memcache works on 64-bit unsigned integers. If decrementing
   * causes the value to become negative it will stay at 0.
   *
   * If the value is negative this will work as if `increment` was called.
   *
   * The updated value is returned.
   */
  Future<int> decrement(key, {int delta: 1, int initialValue: 0});

  /**
   * Delete all items in memcache.
   */
  Future clear();

  /**
   * Returns a new instance of `Memcache` which keeps track of the
   * Data Version Check, also called 'Compare And Set' (CAS).
   *
   * When using this instance each `get` and `getAll` operation will retreive
   * the CAS value and keep track of that.
   *
   * When a `set` or `setAll` operation is performed, the CAS value retrieved
   * from the `get` or `getAll` operation will be used, to ensure that setting
   * a value only happens if the value has not been modified since it was
   * retreived.
   *
   * The CAS check is only in effect when `set` or `setAll` uses the action
   * `SetAction.SET`. This is the default action.
   *
   * For the other operations `remove`, `increment` and `decrement` the CAS
   * value is not used and these operations will be preformed unconditialally.
   *
   * NOTE: There is no way of clearing the tracked CAS values. Just create a
   * new `withCAS` instance when needed.
   */
  Memcache withCAS();
}

class SetAction {
  final int _action;
  final String _name;
  static const SetAction SET = const SetAction._(0, "SET");
  static const SetAction ADD = const SetAction._(1, "ADD");
  static const SetAction REPLACE = const SetAction._(2, "REPLACE");

  const SetAction._(this._action, this._name);

  int get hashCode => _action.hashCode;

  String toString() => 'SetAction($_name)';
}


lib/src/memcache_impl.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library memcache.impl;

import 'dart:async';
import 'dart:collection';
import 'dart:convert';

import '../memcache.dart';
import '../memcache_raw.dart' as raw;

class MemCacheImpl implements Memcache {
  final raw.RawMemcache _raw;
  final bool _withCas;
  Map<List<int>, int> _cas;
  Expando _hashCache;

  MemCacheImpl(this._raw, {withCas: false}) : _withCas = withCas {
    if (withCas) {
      _cas = new HashMap(equals: _keyCompare, hashCode: _keyHash);
      _hashCache = new Expando<int>();
    }
  }

  // Jenkins's one-at-a-time hash, see:
  // http://en.wikipedia.org/wiki/Jenkins_hash_function.
  int _keyHash(List<int> key) {
    final MASK_32 = 0xffffffff;
    int add32(x, y) => (x + y) & MASK_32;
    int shl32(x, y) => (x << y) & MASK_32;

    var cached = _hashCache[key];
    if (cached != null) {
      return cached;
    }

    int hash = 0;
    for (int i = 0; i < key.length; i++) {
      hash = add32(hash, key[i]);
      hash = add32(hash, shl32(hash, 10));
      hash ^= hash >> 6;
    }

    hash = add32(hash, shl32(hash, 3));
    hash ^= hash >> 11;
    hash = add32(hash, shl32(hash, 15));
    _hashCache[key] = hash;
    return hash;
  }

  bool _keyCompare(List<int> key1, List<int> key2) {
    if (key1.length != key2.length) return false;
    for (int i = 0; i < key1.length; i++) {
      if (key1[i] != key2[i]) return false;
    }
    return true;
  }

  List<int> _createKey(Object key) {
    if (key is String) {
      key = utf8.encode(key);
    } else {
      if (key is! List<int>) {
        throw new ArgumentError('Key must have type String or List<int>');
      }
    }
    return key;
  }

  List<int> _createValue(Object value) {
    if (value is String) {
      value = utf8.encode(value);
    } else {
      if (value is! List<int>) {
        throw new ArgumentError('Value must have type String or List<int>');
      }
    }
    return value;
  }

  void _checkExpiration(Duration expiration) {
    const int secondsInThirtyDays = 60 * 60 * 24 * 30;
    // Expiration cannot exceed 30 days.
    if (expiration != null && expiration.inSeconds > secondsInThirtyDays) {
      throw new ArgumentError('Expiration cannot exceed 30 days');
    }
  }

  raw.GetOperation _createGetOperation(List<int> key) {
    return new raw.GetOperation(key);
  }

  raw.SetOperation _createSetOperation(
      Object key, Object value, SetAction action, Duration expiration) {
    var operation;
    var cas;
    switch (action) {
      case SetAction.SET:
        operation = raw.SetOperation.SET;
        if (_withCas) cas = _findCas(key);
        break;
      case SetAction.ADD:
        operation = raw.SetOperation.ADD;
        break;
      case SetAction.REPLACE:
        operation = raw.SetOperation.REPLACE;
        break;
      default:
        throw new ArgumentError('Unsupported set action $action');
    }
    var exp = expiration != null ? expiration.inSeconds : 0;
    return new raw.SetOperation(
        operation, key, 0, cas, _createValue(value), exp);
  }

  raw.RemoveOperation _createRemoveOperation(Object key) {
    return new raw.RemoveOperation(_createKey(key));
  }

  raw.IncrementOperation _createIncrementOperation(
      Object key, int direction, int delta, int initialValue) {
    if (delta is! int) {
      throw new ArgumentError('Delta value must have type int');
    }
    return new raw.IncrementOperation(
        _createKey(key), delta, direction, 0, initialValue);
  }

  void _addCas(List<int> key, int cas) {
    _cas[key] = cas;
  }

  int _findCas(List<int> key) {
    return _cas[key];
  }

  Future get(Object key, {bool asBinary: false}) {
    return new Future.sync(
            () => _raw.get([_createGetOperation(_createKey(key))]))
        .then((List<raw.GetResult> response) {
      if (response.length != 1) {
        throw const MemcacheError.internalError();
      }
      var result = response.first;
      if (_withCas) {
        _addCas(key, result.cas);
      }
      if (result.status == raw.Status.KEY_NOT_FOUND) return null;
      if (result.status == raw.Status.NO_ERROR) {
        return asBinary ? result.value : utf8.decode(result.value);
      }
      throw new MemcacheError(result.status, 'Error getting item');
    });
  }

  Future<Map> getAll(Iterable keys, {bool asBinary: false}) {
    return new Future.sync(() {
      // Copy the keys as they might get mutated by _createKey below.
      var keysList = keys.toList();
      var binaryKeys = new List(keysList.length);
      var request = new List<raw.GetOperation>(keysList.length);
      for (int i = 0; i < keysList.length; i++) {
        binaryKeys[i] = _createKey(keysList[i]);
        request[i] = _createGetOperation(binaryKeys[i]);
      }
      return _raw.get(request).then((List<raw.GetResult> response) {
        if (response.length != request.length) {
          throw const MemcacheError.internalError();
        }
        var result = new Map();
        for (int i = 0; i < keysList.length; i++) {
          var value;
          final responseStatus = response[i].status;
          if (responseStatus == raw.Status.KEY_NOT_FOUND) {
            value = null;
          } else if (responseStatus == raw.Status.NO_ERROR) {
            value =
                asBinary ? response[i].value : utf8.decode(response[i].value);
            if (_withCas) {
              _addCas(binaryKeys[i], response[i].cas);
            }
          } else {
            throw new MemcacheError(responseStatus, 'Error getting item');
          }
          result[keysList[i]] = value;
        }

        return result;
      });
    });
  }

  Future set(key, value,
      {Duration expiration, SetAction action: SetAction.SET}) {
    return new Future.sync(() {
      _checkExpiration(expiration);
      key = _createKey(key);
      return _raw
          .set([_createSetOperation(key, value, action, expiration)]).then(
              (List<raw.SetResult> response) {
        if (response.length != 1) {
          // TODO(sgjesse): Improve error.
          throw const MemcacheError.internalError();
        }
        var result = response.first;
        if (result.status == raw.Status.NO_ERROR) return null;
        if (result.status == raw.Status.NOT_STORED) {
          throw const NotStoredError();
        }
        if (result.status == raw.Status.KEY_EXISTS) {
          throw const ModifiedError();
        }
        throw new MemcacheError(result.status, 'Error storing item');
      });
    });
  }

  Future setAll(Map keysAndValues,
      {Duration expiration, SetAction action: SetAction.SET}) {
    return new Future.sync(() {
      _checkExpiration(expiration);
      var request = <raw.SetOperation>[];
      keysAndValues.forEach((key, value) {
        key = _createKey(key);
        request.add(_createSetOperation(key, value, action, expiration));
      });
      return _raw.set(request).then((List<raw.SetResult> response) {
        if (response.length != request.length) {
          throw const MemcacheError.internalError();
        }
        response.forEach((raw.SetResult result) {
          if (result.status == raw.Status.NO_ERROR) return;
          if (result.status == raw.Status.NOT_STORED) {
            // If one element is not stored throw NotStored.
            throw const NotStoredError();
          }
          if (result.status == raw.Status.KEY_EXISTS) {
            // If one element is modified throw Modified.
            throw const ModifiedError();
          }
          // If one element has another status throw.
          throw new MemcacheError(result.status, 'Error storing item');
        });
        return null;
      });
    });
  }

  Future remove(key) {
    return new Future.sync(() => _raw.remove([_createRemoveOperation(key)]))
        .then((List<raw.RemoveResult> responses) {
      final result = responses[0];
      if (result.status != raw.Status.NO_ERROR &&
          result.status != raw.Status.KEY_NOT_FOUND) {
        throw new MemcacheError(result.status, 'Error removing item');
      }

      // The remove is considered succesful no matter whether the key was
      // there or not.
      return null;
    });
  }

  Future removeAll(Iterable keys) {
    return new Future.sync(() {
      var request = keys.map(_createRemoveOperation).toList(growable: false);
      return _raw.remove(request).then((List<raw.RemoveResult> responses) {
        if (responses.length != request.length) {
          throw const MemcacheError.internalError();
        }

        for (final result in responses) {
          if (result.status != raw.Status.NO_ERROR &&
              result.status != raw.Status.KEY_NOT_FOUND) {
            throw new MemcacheError(result.status, 'Error removing item');
          }
        }

        // The remove is considered succesful no matter whether the key was
        // there or not.
        return null;
      });
    });
  }

  Future<int> increment(key, {int delta: 1, int initialValue: 0}) {
    var direction = delta >= 0
        ? raw.IncrementOperation.INCREMENT
        : raw.IncrementOperation.DECREMENT;
    return new Future.sync(() => _raw.increment([
          _createIncrementOperation(key, direction, delta.abs(), initialValue)
        ])).then((List<raw.IncrementResult> responses) {
      if (responses.length != 1) {
        // TODO(sgjesse): Improve error.
        throw const MemcacheError.internalError();
      }
      var response = responses[0];
      if (response.status != raw.Status.NO_ERROR) {
        throw new MemcacheError(response.status, response.message);
      }
      return response.value;
    });
  }

  Future<int> decrement(key, {int delta: 1, int initialValue: 0}) {
    return increment(key, delta: -delta, initialValue: initialValue);
  }

  Future clear() {
    return new Future.sync(() => _raw.clear());
  }

  Memcache withCAS() {
    return new MemCacheImpl(_raw, withCas: true);
  }
}


lib/memcache_raw.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library memcache.raw;

import "dart:async";

import 'src/dialer.dart' show Dialer;
import 'src/memcache_native_connection.dart' as api;

class Status {
  final int _status;
  final String _name;
  static Status NO_ERROR = const Status._(0x0000, 'NO_ERROR');
  static Status KEY_NOT_FOUND = const Status._(0x0001, 'KEY_NOT_FOUND');
  static Status KEY_EXISTS = const Status._(0x0002, 'KEY_EXISTS');
  static Status NOT_STORED = const Status._(0x0005, 'NOT_STORED');
  static Status ERROR = const Status._(0x0084, 'ERROR');

  const Status._(this._status, this._name);

  int get hashCode => _status.hashCode;

  String toString() => 'Status($_name)';
}

/**
 * Low level memcache interface providing access to all details.
 */
abstract class RawMemcache {
  Future<List<GetResult>> get(List<GetOperation> batch);
  Future<List<SetResult>> set(List<SetOperation> batch);
  Future<List<RemoveResult>> remove(List<RemoveOperation> batch);
  Future<List<IncrementResult>> increment(List<IncrementOperation> batch);
  Future clear();
}

class GetOperation {
  final List<int> key;

  GetOperation(this.key);

  String toString() => 'GetOperation(key: $key)';
}

class GetResult {
  final Status status;
  final String message;
  final int flags;
  final int cas;
  final List<int> value;

  GetResult(this.status, this.message, this.flags, this.cas, this.value);

  String toString() =>
      'GetResult(status: $status, message: $message, flags: $flags, '
      'cas: $cas, value: $value)';
}

class SetOperation {
  static const int SET = 0;
  static const int ADD = 1;
  static const int REPLACE = 2;
  static const int CAS = 3;

  final int operation;
  final List<int> key;
  final int flags;
  final int cas;
  final List<int> value;
  // Expiration time in seconds. If this value is less than the number of
  // seconds in 30 days (60 * 60 * 24 * 30) the expiration is relative to the
  // current time. If the value is higher than the number of seconds in 30 days
  // the expiration is absolute as seconds since the epoc.
  //
  // See https://github.com/memcached/memcached/blob/master/doc/protocol.txt.
  final int expiration;

  SetOperation(this.operation, this.key, this.flags, this.cas, this.value,
      this.expiration);

  String toString() =>
      'SetOperation(operation: $operation, key: $key, flags: $flags, '
      'cas: $cas, value: $value, expiration: $expiration)';
}

class SetResult {
  final Status status;
  final String message;

  SetResult(this.status, this.message);

  String toString() => 'SetResult(status: $status, message: $message)';
}

class RemoveOperation {
  final List<int> key;

  RemoveOperation(this.key);

  String toString() => 'RemoveOperation(key: $key)';
}

class RemoveResult {
  final Status status;
  final String message;

  RemoveResult(this.status, this.message);

  String toString() => 'RemoveResult(status: $status, message: $message)';
}

class IncrementOperation {
  static const int INCREMENT = 0;
  static const int DECREMENT = 1;

  final List<int> key;
  final int delta;
  final int direction;
  final int initialFlags;
  final int initialValue;

  IncrementOperation(this.key, this.delta, this.direction, this.initialFlags,
      this.initialValue);

  String toString() =>
      'IncrementOperation(key: $key, delta: $delta, direction: $direction, '
      'initialFlags: $initialFlags, initialValue: $initialValue)';
}

class IncrementResult {
  final Status status;
  final String message;
  final int value;

  IncrementResult(this.status, this.message, this.value);

  String toString() => 'IncrementResult(status: $status, message: $message)';
}

/// Implements the binary memcached protocol and uses a [Dialer] to (re)connect
/// to the server.
class BinaryMemcacheProtocol implements RawMemcache {
  final Dialer _dialer;
  api.MemCacheNativeConnection _connection;

  BinaryMemcacheProtocol._(this._dialer);

  factory BinaryMemcacheProtocol(String host, int port) {
    return new BinaryMemcacheProtocol._(new Dialer(host, port));
  }

  Future<List<GetResult>> get(List<GetOperation> batch) async {
    if (_connection == null || _connection.isClosed) {
      _connection = await _dialer.dial();
    }
    final futures = batch.map((GetOperation op) {
      final request = new api.Request.get(op.key);
      return _connection.sendRequest(request);
    });

    final List<api.Response> responses = await Future.wait(futures);
    return responses.map((api.Response response) {
      switch (response.status) {
        case api.ResponseStatus.KEY_NOT_FOUND:
          return new GetResult(Status.KEY_NOT_FOUND, response.statusMessage,
              0 /*response.flags*/, response.cas, response.value);
        case api.ResponseStatus.NO_ERROR:
          return new GetResult(Status.NO_ERROR, null, 0 /*response.flags*/,
              response.cas, response.value);
        default:
          return new GetResult(Status.ERROR, response.statusMessage,
              0 /*response.flags*/, response.cas, response.value);
      }
    }).toList(growable: false);
  }

  Future<List<SetResult>> set(List<SetOperation> batch) async {
    if (_connection == null || _connection.isClosed) {
      _connection = await _dialer.dial();
    }
    final futures = batch.map((SetOperation op) {
      switch (op.operation) {
        case SetOperation.SET:
          return _connection.sendRequest(new api.Request.set(op.key, op.value,
              flags: op.flags ?? 0,
              cas: op.cas ?? 0,
              expiration: op.expiration ?? 0));
        case SetOperation.ADD:
          return _connection.sendRequest(new api.Request.add(op.key, op.value,
              flags: op.flags ?? 0, expiration: op.expiration ?? 0));
        case SetOperation.REPLACE:
          return _connection.sendRequest(new api.Request.replace(
              op.key, op.value,
              flags: op.flags ?? 0,
              cas: op.cas ?? 0,
              expiration: op.expiration ?? 0));
        case SetOperation.CAS:
          return _connection.sendRequest(new api.Request.set(op.key, op.value,
              flags: op.flags ?? 0,
              cas: op.cas ?? 0,
              expiration: op.expiration ?? 0));
      }
    });

    final List<api.Response> responses = await Future.wait(futures);
    return responses.map((api.Response response) {
      switch (response.status) {
        case api.ResponseStatus.NO_ERROR:
          return new SetResult(Status.NO_ERROR, null);
        case api.ResponseStatus.ITEM_NOT_STORED:
        case api.ResponseStatus.KEY_NOT_FOUND:
          return new SetResult(Status.NOT_STORED, response.statusMessage);
        case api.ResponseStatus.KEY_EXISTS:
          return new SetResult(Status.NOT_STORED, response.statusMessage);
        default:
          return new SetResult(Status.ERROR, response.statusMessage);
      }
    }).toList(growable: false);
  }

  Future<List<RemoveResult>> remove(List<RemoveOperation> batch) async {
    if (_connection == null || _connection.isClosed) {
      _connection = await _dialer.dial();
    }
    final futures = batch.map((RemoveOperation op) {
      final request = new api.Request.delete(op.key);
      return _connection.sendRequest(request);
    });

    final List<api.Response> responses = await Future.wait(futures);
    return responses.map((api.Response response) {
      switch (response.status) {
        case api.ResponseStatus.NO_ERROR:
          return new RemoveResult(Status.NO_ERROR, null);
        case api.ResponseStatus.KEY_NOT_FOUND:
          return new RemoveResult(Status.KEY_NOT_FOUND, null);
        default:
          return new RemoveResult(Status.ERROR, response.statusMessage);
      }
    }).toList(growable: false);
  }

  Future<List<IncrementResult>> increment(
      List<IncrementOperation> batch) async {
    if (_connection == null || _connection.isClosed) {
      _connection = await _dialer.dial();
    }
    final futures = batch.map((IncrementOperation op) {
      switch (op.direction) {
        case IncrementOperation.INCREMENT:
          return _connection.sendRequest(
              new api.Request.increment(op.key, op.delta, op.initialValue));
        case IncrementOperation.DECREMENT:
          return _connection.sendRequest(
              new api.Request.decrement(op.key, op.delta, op.initialValue));
        default:
          throw new ArgumentError(
              'The direction must be increment or decrement!');
      }
    });

    final List<api.Response> responses = await Future.wait(futures);
    return responses.map((api.Response response) {
      switch (response.status) {
        case api.ResponseStatus.NO_ERROR:
          return new IncrementResult(
              Status.NO_ERROR, null, response.incrDecrValue);
        case api.ResponseStatus.KEY_NOT_FOUND:
          return new IncrementResult(
              Status.KEY_NOT_FOUND, response.statusMessage, 0);
        default:
          return new IncrementResult(Status.ERROR, response.statusMessage, 0);
      }
    }).toList(growable: false);
  }

  Future clear() async {
    if (_connection == null || _connection.isClosed) {
      _connection = await _dialer.dial();
    }

    final api.Response response =
        await _connection.sendRequest(new api.Request.flush());
    if (response.status != api.ResponseStatus.NO_ERROR) {
      throw new Exception(
          'Failed to clear memcache (which should never happen)');
    }
  }

  Future close() => _connection?.close();
}
